PM: Fix deadlock and documentation enhancements

ceolin opened the pull request:
merged https://github.com/zephyrproject-rtos/zephyr/pull/35315 and https://github.com/zephyrproject-rtos/zephyr/pull/35180

Commit:
docs: pm: Add device runtime definition

Add a brief definition about device runtime.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Fix typo

s/Enbles/Enables/g

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Fix doxygen tag

Uses tag "option" for CONFIG_PM_DEVICE_RUNTIME to make it
linkable. Just like is done for all others build options.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Fix invalid API reference

s/device_set_power_state/pm_device_state_set/g

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Device pm fixes and enhancements

- idle/runtime
- Information about asynchronous calls and how to use the condition
  variable
- More details about usage count

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Add missing device states

Adding documentation about two new device states, SUSPENDING and
RESUMING.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Fix device busy information

Remove invalid information about device busy API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Add link to application defined policy

Add a link for the sample that defines its own policy manager.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Improve residency policy documentation

Add information about exit latency time.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Add samples and tests links

Add some applications samples to the documentation.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Fix enable / disable runtime pm

Device pm runtime should only be enabled if device supports pm.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Fix atomic usage

Protect critical sessions using the spinlock available. The atomic
usage was not properly protecting the critical section and was
possible to have a race condition between the usage check and state
set.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Add helper to wait on async ops

Add a function that properly uses a mutex to check a condition before
wait on the conditional variable.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: pm: Add device_runtime test

- Test two threads changing states concurrently
- Test multiple calls to get/put
- Test async / sync API

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Fix sync path

Properly wait on conditional variable using the new API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: device_pm: Fix condvar usage

Use pm_device_wait() to properly wait an operation to finish.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: get rid of the spinlock

Protect critical sections using the mutex.
The mutex is required to use the conditional variable and since we
need to atomically check the pm state and the workqueue before wait
the condition, it is necessary to protect them using the same mutex.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Document isr and pre-kernel functions

Add notes for APIs that are allowed to be called from ISRs and
pre-kernel.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: device_pm: Fix harness

Changing from spinlock to mutex makes the running thread be preempted
and the workqueue ran. This changed the prints order in this sample.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Get rid of atomic for state

Since we are using mutex to protect critical sections and mutexes are
reentrant, it is possible to get rid of atomic for the state because
we can lock the mutex in device_pm_callback.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Don't use k_work on synchronous calls

Do not use worqueue on put/get synchronous calls.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Change API behavior s/_sync/_async

Most APIs have the default synchronous and an asynchronous version
with the sufix _async because that is the most common use.

All devices in tree right now are using the synchronous version, so
just change it to be consistent with the rest of the system.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
```
Setting the state should be based on the condition variable, so this line really needs to be after checking the condition.

ceolin replied:
it does not to be fair. The state and the queue, that governs async ops, are protected by the mutex and not by the conditional variable. This code assumes that the device can do it asynchronously and because of that we need to wait after the operation (because only from the callback signal we will know that the operation finished). But, I saw the next comment and I got that this is coming from the assumption that devices do it synchronously.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
+	(void)k_mutex_unlock(&dev->pm->lock);
```
the unlock needs to be after calling wait(). You are mixing locking via a mutex with condition variables here.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
+	(void)k_mutex_unlock(&dev->pm->lock);
+	pm_device_wait(dev, K_FOREVER);
```
I would expand the contents of this function here less the mutex lock/unlock which we already have in place here.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
 	 * thread we check it here before returning.
 	 */
-	ret = target_state == atomic_get(&dev->pm->state) ? 0 : -EIO;
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
```
there will not be a need for this additional lock, we are locked above, so basically it should end up like this:

```

	while ((k_work_delayable_is_pending(&dev->pm->work)) ||
		(dev->pm->state == PM_DEVICE_STATE_SUSPENDING) ||
		(dev->pm->state == PM_DEVICE_STATE_RESUMING)) {
		ret = k_condvar_wait(&dev->pm->condvar, &dev->pm->lock,
			       K_FOREVER);
		if (ret != 0) {
			break;
		}
	}
	pm_device_runtime_state_set(dev->pm);
	/*
	 * dev->pm->state was set in device_pm_callback(). As the device
	 * may not have been properly changed to the target_state or another
	 * thread we check it here before returning.
	 */
	ret = target_state == dev->pm->state ? 0 : -EIO;

out_unlock:
	(void)k_mutex_unlock(&dev->pm->lock);
```


The callback will be called synchronously, so the above should work.

ceolin replied:
Ok, let be very clear. This **just** works because we are assuming that the device is doing it synchronous. That is a condition that  we are putting on devices

ceolin replied:
if the device does it asynchronous we may return before the callback be called with a transitional state, RESUMING or SUSPENDING.

nashif replied:
async calls do not get here, they are now being handled by the workqueue. This is something that can be documented, a device driver might also choose not to call the callback at all, which another pitfall.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
-	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+
+	if (!dev->pm->enable) {
```
why does this need to be called with the lock? We could just move the lock below this block.

ceolin replied:
nope, if a different thread disables runtime pm ?

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
```
can you add some comments on this condition? When would the usage be negative after incrementing it?

ceolin replied:
That is just because the old implementation was allowing the reference count gets negative due to unbalanced pm_device_put() calls. I just kept this behavior for now.

ceolin replied:
I think it is fine removing it here though. Just did it.

erwango replied:
I'm not clear on this part neither. Do you plan an update ?

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
```
In the synchronous case, what are we increasing usage here and not after checking the condition below and at the same time we are changing the state..?

ceolin replied:
which check below you are referring ?

nashif replied:
before calling `pm_device_runtime_state_set`

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -24,7 +24,9 @@ static void device_pm_callback(const struct device *dev,
-	atomic_set(&dev->pm->state, *state);
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+	dev->pm->state = *state;
+	(void)k_mutex_unlock(&dev->pm->lock);
```
unlock after broadcast.

ceolin replied:
ok, assuming that this callback will be called by the device in `pm_device_state_set`, the mutex will be locked and not necessary here.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
```
everything here is going to be run synchronously, so we do not need to lock/unlock the mutex in both `pm_device_runtime_state_set` and the callback which will be covered by the existing mutex that was taken above.

ceolin replied:
pm_device_runtime_state_set is also used by the workqueue, that is why there is a mutex there, as mutex are re-entrant, it is not a problem lock() twice from the same thread (sync path).  Alos,v the callback was not assuming that would run synchronous. Again, in the end all requested changes are based on the assumption that devices do it synchronously.

nashif replied:
agree about mutexes holding a count, but it is still another call that needs to be made which we could save. One more important point to note when we get to async, we can't use mutexes in async calls, mutexes can't be unlocked in ISR context.

Commit:
pm: device_runtime: Simplify mutex usage

Assuming that pm_device_state_set is synchronous it is possible to
simplify the mutex usage. Now there are two places where the lock is
held, one in the worqueue handler and other in pm_device_request to
cover the synchronous path. It is no longer needed held the lock in the
pm_device_state_set callback and not needed to wait on the conditional
variable after set the state in the synchronous path.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Remove not necessary check

pm_device_runtime_state_set takes care of the check the reference count
and take the right action. It is not necessary check it in
pm_device_request.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@nashif  thanks for reviewing it.I think I've addressed the comments in the last two commits that I put on top of this pr. Can you look it again ?
